Add jetbrains IDEs

syndbg opened the pull request:
WIP on issue #60 .

Added:
  - [x] **RubyMine** in #71 (which was moved here)  
  - [x] **PyCharm Professional**
  - [x] **Pycharm Educational**
  - [x] **WebStorm**
  - [x] **PHPStorm**
  - [ ] **CLion**

didrocks commented:
Excellent work, looking awesome!

At first sight, everything looks good (and thanks for the pythonish-style fixes in umake/tools.py as well!). You added the medium tests and everything and I didn't spot any obvious mistake.

I'm on a slow network until Monday, so can't really run large tests nor medium one to build the container, but will definitively try and give more feedback by then.

Thanks again for having so quickly proposed those 3 new IDEs support!

Tinche commented on tests/data/server-content/www.jetbrains.com/pycharm/download/download_thanks.jsp?edition=edu&os=linux:
```diff
@@ -0,0 +1,408 @@
+
+
+        <h2>Thank you for downloading PyCharm Educational Edition!</h2>
+        <h4 class="sans_serif_normal">Your download should start shortly. If it doesn't, please use one of these direct links: <a href="http://download.jetbrains.com/python/pycharm-educational-fake.tar.gz">HTTP</a> or <a href="https://download.jetbrains.com/python/pycharm-educational-1.0.1.tar.gz">HTTPS</a>.</h4>
```
You're replaced the wrong link here so the medium test throws an error (404). Use the HTTPS link, not the HTTP.

Tinche commented:
Great work, Anton! Alongside the comment I left inline, the java-fake script inside pycharm-professional-fake.tar.gz isn't executable (chmod u+x) so the test won't be able to succeed. Just repack it with the execute bit set :)

syndbg commented:
Addressed the issues. :8ball: 

You're a great help @Tinche :+1: 

@didrocks Aaah, you're too kind. It's all thanks to @Tinche ! 

Btw do you have an IRC channel where ideas/suggestions can be discussed?

Tinche commented:
PyCharm Professional medium test works (and works in general, using it now :)

PyCharm Educational medium test fails - the code looks for the page at https://www.jetbrains.com/pycharm-educational/download/download_thanks.jsp?os=linux, but the test server inside the container will serve it at pycharm/download/download_thanks.jsp?edition=edu&os=linux. The file just needs to be moved to the proper directory (pycharm-educational).

Tinche commented:
There's another problem, but not in the implementation, we just need to figure out what to do.

When we install these products, we create .desktop files for them - which is as it should be, because how else are you going to run the software for the first time? But the products themselves can also create .desktop files for themselves, and offer to do so when first started, in some kind of welcome wizard. There is also an item in the 'Tools' menu - 'Create Desktop Entry' - which does the same thing.

The problem here is that we should create the same file as the products themselves so the products overwrite them, or users will end up with two IDEA/PyCharm/RubyMine/etc items in the dash (like I do right now). However, as far as I can see, both PyCharm Community Edition and PyCharm Professional edition use the same .desktop filename - 'jetbrains-pycharm.desktop'. (Someone should check for the Educational edition too. I'll check IDEA.) I guess JetBrains don't count on users having the same product in multiple editions installed concurrently.

The best thing short-term would probably be to do as JetBrains do; for all our PyCharm editions to create the same .desktop files - 'jetbrains-pycharm.desktop' - and assume our users won't try installing multiple editions. Long term we should probably ask the JetBrains folks to use different .desktop filenames for their products.

syndbg commented:
Fair! I won't lie, I tried running the tests but after the at least an hour of waiting for the tests to complete, I gave up :hammer: 

However, it's 11pm here, so I'll look into these tomorrow :)

Tinche commented:
No problem, nobody's in a hurry. :) You can actually run the tests one by one like so:

```
./runtests --debug tests.medium.test_ide.PyCharmEducationalIDEInContainer
```

didrocks commented:
@Tinche: thanks for doing the review! I'll let you finishing it up as you dive into it more than I, do not hesitate if you need any help :)

I agree with your proposal, let's do as the jetbrain guys are doing, having the same desktop file name for now and assume our users won't try to install multiple of them.

Or, we can maybe install a new one and create a symlink (so that if someone install pycharm and community edition and professional, they still have launcher icons)? The issue is that the Unity dash and other tools will show the duplicates, but I can patch out the dash in vivid for this. Anyway, let's keep it simple now, and we should contact the jetbrain guys.

didrocks commented:
Hey guys!

FYI, I'm about to cut a release (probably tomorrow) and another one next week before ubuntu vivid feature freeze. I was wondering if you need anything from me on the pull request so that we can get it in the second release? I'm happy to help/fix things where you are stuck :)

Keep me posted!

syndbg commented:
I should finally give a status update.

Currently only CLion is left.
Due to life obstacles I might not finish it on time.

Can you review the PR one more time and see if it's good to merge without CLion?

didrocks commented:
Sorry, I was away for a long week-end until today.

Sounds good to me! I'll try to rebase on trunk (seems you can't merge your proposal on trunk) and will do the review today! Good luck and wish all the best for your life :)

didrocks commented:
Done! I had to remerge and reapply manually each commit to rebase on master. I also fixed most of medium tests which didn't work, but I agree that it's not obvious on how to run them. Feel free to ask any questions for future contributions, I'm happy to help/document how to work with them in more details.

I did add some commits (looks at master if you are interested) to reorganize logically the commits while preserving your authoring. :)

I also fixed some pep8 issues (./runtests pep8) introduced by some of the commits.

Finally, for future contributions, please do not mix in the same merge pull request features and python/other unrelated changes, that will make easier for us to check and pick what's really needed.

Anyway, that's an awesome contribution (particuarly as it's your first one to this project), and thanks again for doing this! I'm going to cut a release tomorrow (before vivid feature freeze) and of course mention your work here. I hope that life will be easier for you anytime soon and we'll welcome any other contributions :)
Thanks again for your great enhancements!

PS: rereading the discussion: we don't have an official IRC channel, but I'm hanging on #ubuntu-desktop on freenode, with "didrocks" as a nickname, feel free to join! :)

Tinche commented:
Thanks for your contribution :)
